crates/ide/src/hover.rs

```diff
@@ -79,43 +75,35 @@ pub struct HoverResult {
 // image::https://user-images.githubusercontent.com/48062697/113020658-b5f98b80-917a-11eb-9f88-3dbc27320c95.gif[]
 pub(crate) fn hover(
     db: &RootDatabase,
-    range: FileRange,
+    FileRange { file_id, range }: FileRange,
     config: &HoverConfig,
 ) -> Option<RangeInfo<HoverResult>> {
     let sema = hir::Semantics::new(db);
-    let file = sema.parse(range.file_id).syntax().clone();
+    let file = sema.parse(file_id).syntax().clone();
 
-    // This means we're hovering over a range.
-    if !range.range.is_empty() {
-        let expr = find_node_at_range::<ast::Expr>(&file, range.range)?;
-        let ty = sema.type_of_expr(&expr)?;
-
-        if ty.is_unknown() {
-            return None;
-        }
-
-        let mut res = HoverResult::default();
-
-        res.markup = if config.markdown() {
-            Markup::fenced_block(&ty.display(db))
-        } else {
-            ty.display(db).to_string().into()
-        };
-
-        return Some(RangeInfo::new(range.range, res));
-    }
+    let offset = if range.is_empty() {
+        range.start()
+    } else {
+        let expr = file.covering_element(range).ancestors().find_map(|it| {
+            match_ast! {
+                match it {
+                    ast::Expr(expr) => Some(Either::Left(expr)),
+                    ast::Pat(pat) => Some(Either::Right(pat)),
+                    _ => None,
+                }
+            }
+        })?;
```

Comment:
This makes me wonder, would it make sense to have a `impl<L, R> AstNode for Either<L, R>` impl? This is not the first time I could've made use of either for this.

Reply:
Not sure -- cast would've be ambiguous: `Either<Expr, Expr>::cast`, `Either<Expr, ExprBin>::cast`.

Reply:
Thats the one thing that makes me reluctant about this as well.